pkg/doc/jose/jws.go

```diff
@@ -0,0 +1,380 @@
+	}
+
+	kStr, ok := kRaw.(string)
+
+	return kStr, ok
+}
+
+// JSONWebSignature defines JSON Web Signature (https://tools.ietf.org/html/rfc7515)
+type JSONWebSignature struct {
+	ProtectedHeaders   Headers
+	UnprotectedHeaders Headers
+	Payload            []byte
+
+	signature []byte
+}
+
+// SignatureVerifier makes verification of JSON Web Signature.
+type SignatureVerifier interface {
+	// Verify verifies JWS based on the signing input.
+	Verify(joseHeaders Headers, payload, signingInput, signature []byte) error
+}
+
+// SignatureVerifierFunc is a function wrapper for SignatureVerifier.
+type SignatureVerifierFunc func(joseHeaders Headers, payload, signingInput, signature []byte) error
+
+// Verify verifies JWS signature.
+func (s SignatureVerifierFunc) Verify(joseHeaders Headers, payload, signingInput, signature []byte) error {
+	return s(joseHeaders, payload, signingInput, signature)
+}
+
+// CompositeAlgSignatureVerifier defines composite signature verifier based on the algorithm
+// taken from JOSE header alg.
+type CompositeAlgSignatureVerifier struct {
+	verifierByAlg map[string]SignatureVerifier
+}
+
+// NewCompositeAlgSignatureVerifier creates a new CompositeAlgSignatureVerifier
+func NewCompositeAlgSignatureVerifier() *CompositeAlgSignatureVerifier {
+	return &CompositeAlgSignatureVerifier{
+		verifierByAlg: make(map[string]SignatureVerifier),
+	}
+}
+
+// AddVerifier adds a new verifier of algorithm.
+// If algorithm verifier is already present, it's get rewritten.
+func (v *CompositeAlgSignatureVerifier) AddVerifier(
+	alg string, verifier SignatureVerifier) *CompositeAlgSignatureVerifier {
+	v.verifierByAlg[alg] = verifier
+	return v
+}
+
+// Verify verifiers JWS signature.
+func (v *CompositeAlgSignatureVerifier) Verify(joseHeaders Headers, payload, signingInput, signature []byte) error {
+	alg, ok := joseHeaders.Algorithm()
+	if !ok {
+		return errors.New("'alg' JOSE header is not present")
+	}
+
+	verifier, ok := v.verifierByAlg[alg]
+	if !ok {
+		return fmt.Errorf("no verifier found for %s algorithm", alg)
+	}
+
+	return verifier.Verify(joseHeaders, payload, signingInput, signature)
+}
+
+// Signer defines JWS Signer interface. It makes signing of data and provides custom JWS headers relevant to the signer.
+type Signer interface {
+	// Sign signs.
+	Sign(data []byte) ([]byte, error)
+
+	// Headers provides JWS headers. "alg" header must be provided (see https://tools.ietf.org/html/rfc7515#section-4.1)
+	Headers() Headers
+}
+
+// NewJWS creates JSON Web Signature.
+func NewJWS(protectedHeaders, unprotectedHeaders Headers, payload []byte) *JSONWebSignature {
+	return &JSONWebSignature{
+		ProtectedHeaders:   protectedHeaders,
+		UnprotectedHeaders: unprotectedHeaders,
+		Payload:            payload,
+	}
+}
+
+// SerializeCompact makes JWS Compact Serialization (https://tools.ietf.org/html/rfc7515#section-7.1)
+func (s *JSONWebSignature) SerializeCompact(signer Signer, detached bool) (string, error) {
+	err := s.sign(signer)
```

Comment:
It's odd that the object is mutated during serialization. Shouldn't this happen prior?

Reply:
Yeah, it makes sense for me. So we will have the following steps:

```
jws := NewJWS(...)

jws.Sign(signer)

jwsCompacted := jws.SerializeCompact(false) // no detach
```

This will also help to support multiple signings (e.g. for JWS JSON Serialization).

Reply:
This is better. I'm trying to better understand why we don't sign during construction.

Reply:
I changed it to the following:

```
jws := NewJWS(..., signer)

jwsCompacted := jws.SerializeCompact(false) // no detach
```